plugins/woocommerce/src/Blocks/Domain/Services/AddressProviderService.php

```diff
@@ -0,0 +1,102 @@
+<?php
+declare( strict_types=1 );
+namespace Automattic\WooCommerce\Blocks\Domain\Services;
+
+use WC_Address_Provider;
+
+/**
+ * Service class for managing address providers.
+ */
+class AddressProviderService {
+
+	/**
+	 * Cached provider class names from the last filter call.
+	 *
+	 * @var string[]
+	 */
+	private $cached_provider_class_names = [];
+
+	/**
+	 * Cached provider instances.
+	 *
+	 * @var WC_Address_Provider[]
+	 */
+	private $cached_providers = [];
+
+	/**
+	 * Get all registered providers.
+	 *
+	 * @return WC_Address_Provider[] array of WC_Address_Providers.
+	 */
+	public function get_registered_providers(): array {
+		/**
+		 * Filter the registered address providers.
+		 *
+		 * @since 10.2.0
+		 * @param WC_Address_Provider[] $providers Array of fully qualified class names that extend WC_Address_Provider.
+		 */
+		$provider_class_names = apply_filters( 'woocommerce_address_providers', [] );
+
+		$logger = wc_get_logger();
+
+		if ( ! is_array( $provider_class_names ) ) {
+			$logger->error(
+				'Invalid return value for woocommerce_address_providers, expected an array of class names.',
+				[
+					'context' => 'address_provider_service',
+				]
+			);
+			return [];
+		}
+
+		// If the class names haven't changed, return the cached instances.
+		if ( $this->cached_provider_class_names === $provider_class_names && ! empty( $this->cached_providers ) ) {
+			return $this->cached_providers;
+		}
+
+		$providers = [];
+
+		foreach ( $provider_class_names as $provider_class_name ) {
```

Comment:
Shall we just `continue;` if `$provider_class_name` is not a string?

Reply:
Thanks for the suggestion, I refactored to fail early and only add the provider if it passes all tests. Also added unit tests for this.